File: scripts/exec_model/operators/joins.py
"""The pandas join backend — build side set once, probe side streamed per the matrix.

Three executors, one per wire node kind: `HashJoin` (`CudfHashJoin`, nine join types),
`CrossJoin` (`CudfCrossJoin`) and `NestedLoopJoin` (`CudfNestedLoopJoin`, Inner and Left).
Which types may stream a probe side, and what each call emits, is `join_types.capability`
— one function, so this backend and the recipe emulation in `recipe_join.py` cannot hold
different opinions about the matrix.

Per call and at finish, by family:

| type | per probe call | at finish |
|---|---|---|
| INNER | matches | nothing |
| RIGHT (build is left) | matches + this batch's unmatched probe rows | nothing |
| RIGHT_SEMI / RIGHT_ANTI | this batch's probe rows with / without a match | nothing |
| LEFT / FULL | matches (+ unmatched probe rows for FULL) | unmatched build rows, null-padded |
| LEFT_SEMI / LEFT_ANTI / LEFT_MARK | nothing | build rows with / without a match, or all plus `mark` |
| CROSS, NESTED LOOP Inner | build × probe batch (predicate-filtered for the NLJ) | nothing |
| NESTED LOOP Left | single-batch probe | — |

The probe-local half of that table is not a convention: with the build side complete, a
probe row that matched nothing in this batch matched nothing anywhere, so its emission is
decided locally. The build-preserving half is the opposite claim, which is why it needs a
finish pass.

**The prototype cheats where the real implementation cannot.** "Which build rows matched at
least once across all probe batches" is kept here as a boolean array over the build frame —
free in-process, and exactly the thing that never crosses the C ABI. The v1 GPU path
accumulates each probe batch's keys and runs one anti/semi join at finish (#136); that is
what `recipe_join.py` models, and running both backends over the same plans is what makes
this file an oracle rather than a second guess.

**Null key equality is explicit**, because pandas and SQL disagree: `merge` matches NaN
keys to each other. With `null_equals_null=False` a row with any null key matches nothing,
implemented by holding those rows out of the merge and treating them as unmatched. Anti and
mark ignore the flag and stay at EQUAL, as the C++ does (#80, #59).
"""

from __future__ import annotations

import numpy as np
import pandas as pd

from ..batch import CallStats
from ..executors import JoinExecutor
from . import join_types as J
from .frame import PandasBatch, concatenate, no_scratch, scratch_of, normalize
from .join_types import JoinType, capability, joined_names, joined_projection

#: suffix the probe side wears inside a merge, so a shared column name cannot collapse two
#: columns into one. The wire keeps both and lets `projection` choose; so does this.
_PROBE_TAG = "__p"
#: output rows / probe rows — the ratio `CardinalityEstimator` returns in gpu_rule.rs.
#: 1.0 is what the constant estimator gives today (#19).
TRIVIAL_FANOUT = 1.0
#: the two row-id marker columns the merged frame carries, int64 each
_MARKER_BYTES_PER_ROW = 16

_BUILD_KEY = "__build_row__"
_PROBE_KEY = "__probe_row__"


def merged_scratch(resident: int, build_rows: int, fanout: float, n_rows: int, n_bytes: int) -> int:
    """Build side (rebuilt per probe call, #136) plus the merged frame.

    The merged frame is sized by the *output* cardinality, which the signature cannot
    derive — but it does not have to. `fanout` is a node property the optimizer supplies at
    plan time and the executor is constructed with, so the estimate is already in `&self`
    by the time this is called. That is what keeps the trait unchanged while the model
    stays correct. Shared with the recipe backend so both model one thing one way.
    """
    if n_rows == 0:
        return resident
    probe_row_bytes = n_bytes / n_rows
    build_row_bytes = resident / max(1, build_rows)
    per_row = probe_row_bytes + build_row_bytes + _MARKER_BYTES_PER_ROW
    return int(resident + n_rows * fanout * per_row)


class _PandasJoin(JoinExecutor):
    """What the three share: the build frame, the residency it is, and the call protocol."""

    def __init__(self, name: str, fanout: float):
        self.name = name
        self.fanout = fanout
        self.build: pd.DataFrame | None = None
        self.probe_calls = 0
        self._probe_columns: list[str] | None = None

    def resident_bytes(self) -> int:
        if self.build is None:
            return 0
        return int(self.build.memory_usage(index=False, deep=True).sum())

    def scratch_bytes(self, n_rows: int, n_bytes: int) -> int:
        build_rows = len(self.build) if self.build is not None else 0
        return merged_scratch(self.resident_bytes(), build_rows, self.fanout, n_rows, n_bytes)

    def _capability(self) -> J.Capability:
        raise NotImplementedError

    def set_build(self, batch: PandasBatch) -> CallStats:
        if self.build is not None:
            raise AssertionError(f"{self.name}: set_build called twice")
        self.build = self._take_build(batch.consume())
        return no_scratch()   # the build frame becomes residency, not scratch

    def _take_build(self, frame: pd.DataFrame) -> pd.DataFrame:
        return frame.copy()

    def probe_and_fetch(self, batch: PandasBatch):
        if self.build is None:
            raise AssertionError(f"{self.name}: probed before set_build")
        self.probe_calls += 1
        if not self._capability().streams and self.probe_calls > 1:
            raise AssertionError(
                f"{self.name}: this join takes a single-batch probe "
                f"({self._capability().reason}); the planner inserts GpuCoalesceAllBatches "
                "under the probe side"
            )
        probe = batch.consume()
        self._probe_columns = list(probe.columns)
        self._transients = []
        frames = self._probe(probe)
        return (
            [PandasBatch(f, f"({self.name}#{self.probe_calls}⋈{batch.tag})") for f in frames],
            scratch_of(*self._transients),
        )

    def finish_and_fetch(self):
        outputs = self._finish()
        # Nothing stays resident after the finish, so the accountant can drop this executor.
        self.build = None
        return [PandasBatch(f, f"({self.name}⋈finish)") for f in outputs], no_scratch()

    def _probe(self, probe: pd.DataFrame) -> list[pd.DataFrame]:
        raise NotImplementedError

    def _finish(self) -> list[pd.DataFrame]:
        return []


class HashJoin(_PandasJoin):
    """`CudfHashJoin`: equi-join on key pairs, with an optional residual filter."""

    def __init__(
        self,
        join_type: JoinType,
        build_keys: list[str],
        probe_keys: list[str],
        null_equals_null: bool = False,
        name: str = "join",
        fanout: float = TRIVIAL_FANOUT,
        probe_schema: list[str] | None = None,
        residual=None,
    ):
        super().__init__(name, fanout)
        if len(build_keys) != len(probe_keys):
            raise ValueError("join key lists must have equal length")
        self.join_type = join_type
        self.build_keys = build_keys
        self.probe_keys = probe_keys
        self.null_equals_null = null_equals_null
        #: the probe side's declared columns — consulted only when an outer finish must
        #: null-pad columns no probe batch ever showed it
        self.probe_schema = probe_schema
        #: a residual (non-equi) predicate over the joined row, as `CudfHashJoin.filter`
        self.residual = residual
        self.matched: np.ndarray | None = None
        #: set when a single-batch probe already emitted everything, finish included
        self._finished = False
        refusal = self._capability().refusal
        if refusal is not None:
            raise NotImplementedError(f"{name}: {refusal}")

    def _capability(self) -> J.Capability:
        return capability(self.join_type, has_filter=self.residual is not None)

    def _take_build(self, frame: pd.DataFrame) -> pd.DataFrame:
        build = frame.copy()
        build[_BUILD_KEY] = np.arange(len(frame))
        self.matched = np.zeros(len(frame), dtype=bool)
        return build

    # -- probe -------------------------------------------------------------------

    def _probe(self, probe: pd.DataFrame) -> list[pd.DataFrame]:
        frames = self._probe_output(probe)
        if self._capability().streams:
            return frames
        # A single-batch probe is the whole join in one call, which is what the legacy
        # path does: nothing waits for a finish, because there is no second batch for it
        # to wait through.
        self._finished = True
        frames += self._finish_frames()
        return [normalize(concatenate(frames))] if len(frames) > 1 else frames

    def _probe_output(self, probe: pd.DataFrame) -> list[pd.DataFrame]:
        pairs = self._matching_pairs(probe)
        self._transients = [pairs]
        if len(pairs):
            self.matched[pairs[_BUILD_KEY].to_numpy()] = True

        if self.join_type in J.FINISH_ONLY:
            return []       # the call exists to update `matched`, nothing else
        if self.join_type in (JoinType.RIGHT_SEMI, JoinType.RIGHT_ANTI):
            hit = probe.index.isin(pairs[_PROBE_KEY].to_numpy())
            keep = hit if self.join_type is JoinType.RIGHT_SEMI else ~hit
            return [normalize(probe[keep])]

        matches = self._joined(pairs, probe.columns)
        if self.join_type in (JoinType.INNER, JoinType.LEFT):
            return [matches]
        # RIGHT / FULL: this batch's unmatched probe rows, build columns null-padded.
        # Batch-local, because the build side is complete before the first probe call.
        unmatched = probe[~probe.index.isin(pairs[_PROBE_KEY].to_numpy())]
        if not len(unmatched):
            return [matches]
        padded = pd.DataFrame(
            {column: np.nan for column in self._build_columns().columns}, index=unmatched.index
        )
        for column in probe.columns:
            padded[f"{column}{_PROBE_TAG}"] = unmatched[column]
        return [normalize(concatenate([matches, self._joined(padded, probe.columns)]))]

    def _matching_pairs(self, probe: pd.DataFrame) -> pd.DataFrame:
        """Every (build row, probe row) the join matches — keys, then the residual.

        The probe side is renamed before the merge so that both sides' columns survive
        even when they share a name. That is what the wire does — a join's output is
        `[left_cols…, right_cols…]` and `TableResult` allows two columns called `k` — and
        it is the join's `projection` that decides which of them the query sees.
        """
        probe = probe.rename(columns={c: f"{c}{_PROBE_TAG}" for c in probe.columns})
        probe[_PROBE_KEY] = np.arange(len(probe))
        probe_keys = [f"{k}{_PROBE_TAG}" for k in self.probe_keys]
        # Rows whose key is null are held OUT of the merge rather than dropped, so an
        # outer type can still emit them as unmatched. That is the whole difference
        # between null_equals_null false and true. Anti and mark are EQUAL regardless.
        equal_nulls = self.null_equals_null or self.join_type in J.IGNORES_NULL_EQUALS_NULL
        if equal_nulls:
            build_live, probe_live = self.build, probe
        else:
            build_live = self.build[self.build[self.build_keys].notna().all(axis=1)]
            probe_live = probe[probe[probe_keys].notna().all(axis=1)]
        pairs = build_live.merge(
            probe_live, how="inner", left_on=self.build_keys, right_on=probe_keys
        )
        if self.residual is not None and len(pairs):
            keep = self.residual.evaluate(self._untagged(pairs, probe.columns))
            pairs = pairs[keep.fillna(False).astype(bool).to_numpy()]
        return pairs

    def _untagged(self, frame: pd.DataFrame, tagged_probe_columns) -> pd.DataFrame:
        """The joined row under its own column names — what a residual filter reads.

        On the wire the filter's `ColumnRef`s index a private intermediate schema and
        `filter_columns` maps each back to a (side, ordinal); here the same mapping is by
        name, per the prototype's expression IR (`expressions.py`).
        """
        renamed = {c: c[: -len(_PROBE_TAG)] for c in tagged_probe_columns if c.endswith(_PROBE_TAG)}
        return frame.rename(columns=renamed)

    def _joined(self, pairs: pd.DataFrame, probe_columns) -> pd.DataFrame:
        """`[build…, probe…]` cut down by the join's `projection`, named as the query sees it.

        The same rule the recipe path states as the node's `projection` field, computed by
        the same function, so the two backends produce one schema rather than two.
        """
        build_names = list(self._build_columns().columns)
        probe_names = list(probe_columns)
        keys = [
            (build_names.index(b), probe_names.index(p))
            for b, p in zip(self.build_keys, self.probe_keys)
        ]
        columns = build_names + [f"{c}{_PROBE_TAG}" for c in probe_names]
        wanted = joined_projection(build_names, probe_names, keys, self.join_type)
        out = pairs[[columns[i] for i in wanted]]
        out.columns = joined_names(build_names, probe_names, keys, self.join_type)
        return normalize(out)

    # -- finish ------------------------------------------------------------------

    def _finish(self) -> list[pd.DataFrame]:
        if self._finished:
            return []
        return self._finish_frames()

    def _finish_frames(self) -> list[pd.DataFrame]:
        """The build rows the probe side never matched — the pass #136 is about."""
        if self.join_type not in J.BUILD_PRESERVING:
            return []
        build = self._build_columns()
        if self.join_type is JoinType.LEFT_SEMI:
            return [normalize(build[self.matched])]
        if self.join_type is JoinType.LEFT_ANTI:
            return [normalize(build[~self.matched])]
        if self.join_type is JoinType.LEFT_MARK:
            return [normalize(build.assign(mark=self.matched))]
        # LEFT / FULL: unmatched build rows, probe columns null-padded.
        out = build[~self.matched]
        for column in self._pad_columns(build):
            out = out.assign(**{column: np.nan})
        return [normalize(out)]

    def _build_columns(self) -> pd.DataFrame:
        return normalize(self.build.drop(columns=[_BUILD_KEY]))

    def _pad_columns(self, build: pd.DataFrame) -> list[str]:
        """The probe columns an outer finish must null-pad.

        Learned from the first probe batch; a probe lane that never produced one falls back
        to the declared `probe_schema`. Without either the output's shape would silently
        depend on whether a probe batch happened to arrive, so the miss is loud.
        """
        probe_columns = self._probe_columns
        if probe_columns is None:
            if self.probe_schema is None:
                raise ValueError(
                    f"{self.name}: a {self.join_type.name} finish saw no probe batch and "
                    "has no probe_schema — the probe columns to null-pad are unknown"
                )
            probe_columns = list(self.probe_schema)
        keys = [
            (list(build.columns).index(b), probe_columns.index(p))
            for b, p in zip(self.build_keys, self.probe_keys)
        ]
        names = joined_names(list(build.columns), probe_columns, keys, self.join_type)
        return [c for c in names if c not in list(build.columns)]


class CrossJoin(_PandasJoin):
    """`CudfCrossJoin` — build × probe batch, one output per call, nothing at finish."""

    def __init__(self, name: str = "cross_join", fanout: float = TRIVIAL_FANOUT):
        super().__init__(name, fanout)

    def _capability(self) -> J.Capability:
        return J.Capability(streams=True, needs_finish=False)

    def _probe(self, probe: pd.DataFrame) -> list[pd.DataFrame]:
        return [normalize(self.build.merge(probe, how="cross"))]


class NestedLoopJoin(_PandasJoin):
    """`CudfNestedLoopJoin` — the cross product filtered by a predicate.

    Inner streams its probe side, as a cross join does. Left does not: its unmatched build
    rows need the finish pass, and #136's trick accumulates *keys*, which a predicate join
    has none of — so the planner gives it a single-batch probe and the whole join happens
    in one call, exactly as the legacy path does it.
    """

    def __init__(
        self,
        join_type: JoinType,
        predicate,
        name: str = "nlj",
        fanout: float = TRIVIAL_FANOUT,
    ):
        super().__init__(name, fanout)
        self.join_type = join_type
        self.predicate = predicate
        refusal = self._capability().refusal
        if refusal is not None:
            raise NotImplementedError(f"{name}: {refusal}")

    def _capability(self) -> J.Capability:
        return capability(self.join_type, has_filter=self.predicate is not None,
                          nested_loop=True)

    def _probe(self, probe: pd.DataFrame) -> list[pd.DataFrame]:
        crossed = self.build.merge(probe, how="cross")
        self._transients = [crossed]
        if self.predicate is None:
            return [normalize(crossed)]
        keep = self.predicate.evaluate(crossed).fillna(False).astype(bool)
        matches = normalize(crossed[keep])
        if self.join_type is JoinType.INNER:
            return [matches]
        # LEFT: the build rows no probe row satisfied, right columns null-padded. One
        # call, so this is not a finish pass — it is the whole join.
        hit = crossed.index.isin(crossed.index[keep])
        matched_build = set(np.repeat(np.arange(len(self.build)), len(probe))[hit].tolist())
        missing = [i for i in range(len(self.build)) if i not in matched_build]
        if not missing:
            return [matches]
        padded = self.build.iloc[missing].copy()
        for column in probe.columns:
            padded[column] = np.nan
        return [normalize(concatenate([matches, padded[list(matches.columns)]]))]

File: peacockdb-core/src/batch_partitioned/nodes/tests/joins.rs
//! The joins: what each of the three requires of its inputs, and the distribution a join
//! declares about its own output — the claim nothing downstream re-checks.

use super::*;

#[test]
fn a_join_whose_build_side_is_many_batches_names_the_node_that_fixes_it() {
    let build = Given::input(one_lane(BatchLayout::MultipleBatches), &["k"]);
    let probe = Given::input(one_lane(BatchLayout::MultipleBatches), &["fk"]);
    let schema = Schema::new(Arc::new(ArrowSchema::new(vec![
        Field::new("k", DataType::Int64, true),
        Field::new("fk", DataType::Int64, true),
    ])));
    let join = GpuCrossJoin::new(build, probe, None, schema);
    invalid(
        join.validate_schemas_and_partitions(),
        "GpuCoalesceAllBatches",
    );
}

#[test]
fn a_join_filter_column_mapped_to_the_wrong_side_is_caught_at_plan_time() {
    use super::join::{JoinFilterColumn, JoinSide};
    let build = Given::input(one_lane(BatchLayout::SingleBatch), &["k"]);
    let probe = Given::input(one_lane(BatchLayout::MultipleBatches), &["fk"]);
    let schema = Schema::new(Arc::new(ArrowSchema::new(vec![
        Field::new("k", DataType::Int64, true),
        Field::new("fk", DataType::Int64, true),
    ])));
    let filter = Expr::binary(
        Expr::column(0, "k"),
        BinaryOp::Lt,
        Expr::column(1, "fk"),
        DataType::Boolean,
    );
    // Both filter columns pointed at the probe: @0 then reads a valid column of the
    // wrong table, which nothing downstream could detect.
    let mapping = vec![
        JoinFilterColumn {
            side: JoinSide::Probe,
            index: 0,
        },
        JoinFilterColumn {
            side: JoinSide::Probe,
            index: 0,
        },
    ];
    let join = GpuNestedLoopJoin::new(
        build,
        probe,
        NestedLoopJoinType::Inner,
        filter,
        mapping,
        None,
        schema,
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "maps to fk on the Probe side",
    );
}

#[test]
fn a_finalize_expression_is_checked_against_the_intermediate_and_not_the_input() {
    // @2 is a column of the input and not of the intermediate table the finalize reads,
    // which is the whole reason the two schemas are separate arguments.
    let input = Given::input(one_lane(BatchLayout::MultipleBatches), &["k", "other", "n"]);
    let intermediate = columns(&["k", "n"]);
    let aggregate = GpuAggregate::new(
        input,
        summing(
            vec![Expr::column(0, "k")],
            vec![Expr::column(2, "n")],
            Some(vec![NamedExpr::new(Expr::column(2, "n"), "n")]),
        ),
        intermediate.clone(),
        columns(&["k", "n"]),
    );
    invalid(
        aggregate.validate_schemas_and_partitions(),
        "GpuAggregate: column n@2 is past the 2 columns",
    );
}

fn hashed(keys: Vec<u32>, n: usize) -> PartitionLayout {
    PartitionLayout {
        key_distribution: KeyDistribution::ByHash { hash_keys: keys },
        batch_layout: BatchLayout::SingleBatch,
        ..PartitionLayout::new(n)
    }
}

fn joining(
    build: Box<dyn GpuNode>,
    probe: Box<dyn GpuNode>,
    join_type: datafusion::common::JoinType,
    filter: Option<Expr>,
    filter_columns: Vec<join::JoinFilterColumn>,
) -> GpuJoin {
    GpuJoin::new(
        build,
        probe,
        join_type,
        vec![(0, 0)],
        filter,
        filter_columns,
        false,
        None,
        columns(&["k", "fk"]),
        join::joined_schema(&columns(&["k"]).fields, &columns(&["fk"]).fields, join_type),
    )
}

#[test]
fn a_join_whose_sides_carry_different_lane_counts_is_refused() {
    let join = joining(
        Given::input(hashed(vec![0], 4), &["k"]),
        Given::input(hashed(vec![0], 8), &["fk"]),
        datafusion::common::JoinType::Inner,
        None,
        Vec::new(),
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "the sides carry 4 and 8 lanes",
    );
}

#[test]
fn a_join_over_lanes_hashed_on_other_columns_names_the_node_that_fixes_it() {
    // Hashed on a column that is not the join key: lane p of one side then holds rows
    // whose matches are in some other lane of the other, and each lane answers for part
    // of a key it cannot see.
    let join = joining(
        Given::input(hashed(vec![1], 4), &["k", "other"]),
        Given::input(hashed(vec![0], 4), &["fk"]),
        datafusion::common::JoinType::Inner,
        None,
        Vec::new(),
    );
    invalid(join.validate_schemas_and_partitions(), "GpuEmitPartitions");
}

#[test]
fn an_equi_join_whose_build_side_is_many_batches_names_the_node_that_fixes_it() {
    let join = joining(
        Given::input(one_lane(BatchLayout::MultipleBatches), &["k"]),
        Given::input(one_lane(BatchLayout::SingleBatch), &["fk"]),
        datafusion::common::JoinType::Inner,
        None,
        Vec::new(),
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "its build side must be one batch per lane",
    );
}

#[test]
fn a_join_that_cannot_stream_its_probe_over_many_batches_names_the_node_that_fixes_it() {
    // A filtered semi join is one legacy call over the whole probe side, so the planner
    // owes it a single batch there — the capability matrix is what says so.
    let join = joining(
        Given::input(one_lane(BatchLayout::SingleBatch), &["k"]),
        Given::input(one_lane(BatchLayout::MultipleBatches), &["fk"]),
        datafusion::common::JoinType::LeftSemi,
        Some(Expr::binary(
            Expr::column(0, "k"),
            BinaryOp::Lt,
            Expr::column(1, "fk"),
            DataType::Boolean,
        )),
        vec![
            join::JoinFilterColumn {
                side: join::JoinSide::Build,
                index: 0,
            },
            join::JoinFilterColumn {
                side: join::JoinSide::Probe,
                index: 0,
            },
        ],
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "cannot stream its probe",
    );
}

#[test]
fn a_join_shape_the_capability_matrix_refuses_is_refused_at_validation() {
    let join = joining(
        Given::input(one_lane(BatchLayout::SingleBatch), &["k"]),
        Given::input(one_lane(BatchLayout::SingleBatch), &["fk"]),
        datafusion::common::JoinType::Left,
        Some(Expr::binary(
            Expr::column(0, "k"),
            BinaryOp::Lt,
            Expr::column(1, "fk"),
            DataType::Boolean,
        )),
        vec![
            join::JoinFilterColumn {
                side: join::JoinSide::Build,
                index: 0,
            },
            join::JoinFilterColumn {
                side: join::JoinSide::Probe,
                index: 0,
            },
        ],
    );
    unsupported(join.validate_schemas_and_partitions(), "#153");
}

#[test]
fn a_join_filter_reference_past_its_map_is_caught_at_plan_time() {
    let join = joining(
        Given::input(one_lane(BatchLayout::SingleBatch), &["k"]),
        Given::input(one_lane(BatchLayout::SingleBatch), &["fk"]),
        datafusion::common::JoinType::Inner,
        Some(Expr::binary(
            Expr::column(0, "k"),
            BinaryOp::Lt,
            Expr::column(1, "fk"),
            DataType::Boolean,
        )),
        vec![join::JoinFilterColumn {
            side: join::JoinSide::Build,
            index: 0,
        }],
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "filter column fk@1 is past the 1 its map has",
    );
}

#[test]
fn a_cross_join_over_several_lanes_names_the_node_that_fixes_it() {
    let join = GpuCrossJoin::new(
        Given::input(hashed(vec![0], 4), &["k"]),
        Given::input(hashed(vec![0], 4), &["fk"]),
        None,
        columns(&["k", "fk"]),
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "both inputs must be one lane, not 4 and 4",
    );
}

#[test]
fn a_left_nested_loop_join_streaming_its_probe_names_the_node_that_fixes_it() {
    // Its finish pass accumulates keys and a predicate join has none, so the probe side
    // has to be whole before the join runs.
    let join = GpuNestedLoopJoin::new(
        Given::input(one_lane(BatchLayout::SingleBatch), &["k"]),
        Given::input(one_lane(BatchLayout::MultipleBatches), &["fk"]),
        NestedLoopJoinType::Left,
        Expr::binary(
            Expr::column(0, "k"),
            BinaryOp::Lt,
            Expr::column(1, "fk"),
            DataType::Boolean,
        ),
        vec![
            join::JoinFilterColumn {
                side: join::JoinSide::Build,
                index: 0,
            },
            join::JoinFilterColumn {
                side: join::JoinSide::Probe,
                index: 0,
            },
        ],
        None,
        columns(&["k", "fk"]),
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "GpuCoalesceAllBatches",
    );
}

#[test]
fn an_interleave_of_branches_with_different_lane_counts_is_refused() {
    let interleave = GpuInterleave::new(
        vec![
            Given::input(hashed(vec![0], 4), &["k"]),
            Given::input(hashed(vec![0], 2), &["k"]),
        ],
        columns(&["k"]),
    );
    invalid(
        interleave.validate_schemas_and_partitions(),
        "same hash distribution",
    );
}

/// The distribution a join declares about its own output, which is the claim nothing
/// downstream re-checks: a co-partitioned join above it is refused or allowed on this.
fn joined_distribution(
    build: (PartitionLayout, &[&str]),
    probe: (PartitionLayout, &[&str]),
    join_type: datafusion::common::JoinType,
    keys: Vec<(u32, u32)>,
    output: &[&str],
) -> KeyDistribution {
    let join = GpuJoin::new(
        Given::input(build.0, build.1),
        Given::input(probe.0, probe.1),
        join_type,
        keys,
        None,
        Vec::new(),
        false,
        None,
        columns(output),
        join::joined_schema(&columns(build.1).fields, &columns(probe.1).fields, join_type),
    );
    input_layout(&join).key_distribution
}

#[test]
fn a_join_over_lanes_nothing_hashed_declares_no_hash() {
    // One lane, no shuffle anywhere below either side: there is no hash, so a claim that
    // rows are where a hash put them is false however the keys are named.
    let unhashed = || (one_lane(BatchLayout::SingleBatch), &["k"][..]);
    assert_eq!(
        joined_distribution(
            unhashed(),
            (one_lane(BatchLayout::MultipleBatches), &["fk"]),
            datafusion::common::JoinType::Inner,
            vec![(0, 0)],
            &["k", "fk"],
        ),
        KeyDistribution::NotSpecified
    );
}

#[test]
fn a_full_join_declares_no_hash_however_its_sides_are_partitioned() {
    // Full pads both ways: an unmatched build row reads NULL where the probe key was and
    // an unmatched probe row reads NULL where the build key was, so neither column holds
    // the value that placed every row.
    assert_eq!(
        joined_distribution(
            (hashed(vec![0], 4), &["k"]),
            (hashed(vec![0], 4), &["fk"]),
            datafusion::common::JoinType::Full,
            vec![(0, 0)],
            &["k", "fk"],
        ),
        KeyDistribution::NotSpecified
    );
}

#[test]
fn a_left_join_names_the_side_whose_rows_are_never_padded() {
    // Its unmatched build rows carry NULL in the probe key, so the build column is the
    // one that holds the placement value for every output row.
    assert_eq!(
        joined_distribution(
            (hashed(vec![0], 4), &["k"]),
            (hashed(vec![0], 4), &["fk"]),
            datafusion::common::JoinType::Left,
            vec![(0, 0)],
            &["k", "fk"],
        ),
        KeyDistribution::ByHash { hash_keys: vec![0] }
    );
}

#[test]
fn a_right_join_names_the_other_one() {
    assert_eq!(
        joined_distribution(
            (hashed(vec![0], 4), &["k"]),
            (hashed(vec![0], 4), &["fk"]),
            datafusion::common::JoinType::Right,
            vec![(0, 0)],
            &["k", "fk"],
        ),
        KeyDistribution::ByHash { hash_keys: vec![1] }
    );
}

#[test]
fn an_inner_join_carries_the_claim_both_its_sides_earned() {
    assert_eq!(
        joined_distribution(
            (hashed(vec![0], 4), &["k"]),
            (hashed(vec![0], 4), &["fk"]),
            datafusion::common::JoinType::Inner,
            vec![(0, 0)],
            &["k", "fk"],
        ),
        KeyDistribution::ByHash { hash_keys: vec![0] }
    );
}

#[test]
fn a_semi_join_carries_the_claim_of_the_side_it_emits() {
    // The output holds the build side alone, at its own ordinals.
    assert_eq!(
        joined_distribution(
            (hashed(vec![1], 4), &["other", "k"]),
            (hashed(vec![0], 4), &["fk"]),
            datafusion::common::JoinType::LeftSemi,
            vec![(1, 0)],
            &["other", "k"],
        ),
        KeyDistribution::ByHash { hash_keys: vec![1] }
    );
}

#[test]
fn sides_hashed_on_different_things_leave_the_output_with_no_claim() {
    // Only one side was scattered on its join key, so the other's rows are not where the
    // first one's are — and an aggregate above would answer per lane for a group that is
    // spread across them.
    assert_eq!(
        joined_distribution(
            (hashed(vec![0], 4), &["k"]),
            (
                PartitionLayout {
                    key_distribution: KeyDistribution::ByHash { hash_keys: vec![1] },
                    batch_layout: BatchLayout::SingleBatch,
                    ..PartitionLayout::new(4)
                },
                &["fk", "other"]
            ),
            datafusion::common::JoinType::Inner,
            vec![(0, 0)],
            &["k", "fk", "other"],
        ),
        KeyDistribution::NotSpecified
    );
}

#[test]
fn a_projection_moves_the_claim_to_where_the_key_actually_lands() {
    // The output ordinal is neither side's: the projection drops a build column, so the
    // build key lands at 0 having been 1 in the crossed table.
    let join = GpuJoin::new(
        Given::input(hashed(vec![1], 4), &["other", "k"]),
        Given::input(hashed(vec![0], 4), &["fk"]),
        datafusion::common::JoinType::Inner,
        vec![(1, 0)],
        None,
        Vec::new(),
        false,
        Some(vec![1, 2]),
        columns(&["k", "fk"]),
        join::joined_schema(
            &columns(&["other", "k"]).fields,
            &columns(&["fk"]).fields,
            datafusion::common::JoinType::Inner,
        ),
    );
    assert_eq!(
        input_layout(&join).key_distribution,
        KeyDistribution::ByHash { hash_keys: vec![0] }
    );
}

#[test]
fn a_key_name_the_output_holds_twice_still_resolves() {
    // Both sides call the key `k`, so a lookup by name has two candidates and can only
    // give up. The ordinal is structural: build width 1, so the probe's key is at 1.
    assert_eq!(
        joined_distribution(
            (hashed(vec![0], 4), &["k"]),
            (hashed(vec![0], 4), &["k"]),
            datafusion::common::JoinType::Right,
            vec![(0, 0)],
            &["k", "k"],
        ),
        KeyDistribution::ByHash { hash_keys: vec![1] }
    );
}

#[test]
fn a_mark_join_carries_the_claim_of_the_side_it_marks() {
    // Its output is the build side plus the boolean, so the build key is still at its own
    // ordinal — a mark join treated as emitting neither side loses a claim it earned, and
    // every co-partitioned node above it then collapses to one lane.
    assert_eq!(
        joined_distribution(
            (hashed(vec![0], 4), &["k"]),
            (hashed(vec![0], 4), &["fk"]),
            datafusion::common::JoinType::LeftMark,
            vec![(0, 0)],
            &["k", "mark"],
        ),
        KeyDistribution::ByHash { hash_keys: vec![0] }
    );
}

#[test]
fn a_semi_joins_projection_is_bounded_by_the_side_it_emits() {
    // A semi join emits the build side alone, so @2 is past its table even though the two
    // sides hold four columns between them — the bound that reads both sides would let it
    // through.
    let join = GpuJoin::new(
        Given::input(one_lane(BatchLayout::SingleBatch), &["k", "other"]),
        Given::input(one_lane(BatchLayout::SingleBatch), &["fk", "spare"]),
        datafusion::common::JoinType::LeftSemi,
        vec![(0, 0)],
        None,
        Vec::new(),
        false,
        Some(vec![2]),
        columns(&["other"]),
        join::joined_schema(
            &columns(&["k", "other"]).fields,
            &columns(&["fk", "spare"]).fields,
            datafusion::common::JoinType::LeftSemi,
        ),
    );
    invalid(
        join.validate_schemas_and_partitions(),
        "projected column @2 is past the 2 a LeftSemi join emits",
    );
}
